adapters/interlocked_hl_win32.c

```diff
@@ -116,6 +117,51 @@ IMPLEMENT_MOCKABLE_FUNCTION(, INTERLOCKED_HL_RESULT, InterlockedHL_WaitForValue,
     return result;
 }
 
+IMPLEMENT_MOCKABLE_FUNCTION(, INTERLOCKED_HL_RESULT, InterlockedHL_WaitForValue64, LONG64 volatile*, address, LONG64, value, DWORD, milliseconds)
+{
+    INTERLOCKED_HL_RESULT result;
+
+    /*Codes_SRS_INTERLOCKED_HL_02_021: [ If address is NULL then InterlockedHL_WaitForValue64 shall fail and return INTERLOCKED_HL_ERROR. ]*/
+    if (address == NULL)
+    {
+        LogError("invalid arguments LONG64 volatile*, address=%p, LONG64, value=%" PRId64 ", DWORD, milliseconds=%" PRIu32 "",
+            address, value, milliseconds);
+        result = INTERLOCKED_HL_ERROR;
+    }
+    else
+    {
+        LONG64 current_value;
+
+        do
+        {
+            /* Codes_SRS_INTERLOCKED_HL_02_025: [ When WaitOnAddress succeeds, the value at address shall be compared to the target value passed in value by using InterlockedAdd64. ]*/
+            current_value = InterlockedAdd64(address, 0);
+            if (current_value == value)
+            {
+                /* Codes_SRS_INTERLOCKED_HL_02_022: [ If the value at address is equal to value then InterlockedHL_WaitForValue64 shall return INTERLOCKED_HL_OK. ]*/
+                result = INTERLOCKED_HL_OK;
+                break;
+            }
+
+            /* Codes_SRS_INTERLOCKED_HL_02_026: [ If the value at address does not match, InterlockedHL_WaitForValue64 shall issue another call to WaitOnAddress. ]*/
+
+            /* Codes_SRS_INTERLOCKED_HL_02_023: [ If the value at address is not equal to value, InterlockedHL_WaitForValue64 shall wait until the value at address changes in order to compare it again to value by using WaitOnAddress. ]*/
+            /* Codes_SRS_INTERLOCKED_HL_02_024: [ When waiting for the value at address to change, the milliseconds argument value shall be used as timeout. ]*/
+            if (!WaitOnAddress(address, &current_value, sizeof(current_value), milliseconds))
+            {
+                LogLastError("failure in WaitOnAddress(address=%p, &current_value=%p, sizeof(current_value)=%zu, milliseconds=%" PRIu32 ")",
```

Comment:
>milliseconds [](start = 113, length = 12)

I think this should be printed as unsigned long. My Intellisense takes me to minwindef.h where it seems to typedef'd to unsigned long.
I know that this MSDN page says it is int32, but honestly unsigned long is not guaranteed to be int32_t... Just saying
[https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types](https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types)

Reply:
Here and below.

---
In reply to: [423407871](https://github.com/Azure/azure-c-util/pull/37#discussion_r423407871) [](ancestors = 423407871)

Reply:
This page [https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress) says:
BOOL WaitOnAddress(
  volatile VOID *Address,
  PVOID         CompareAddress,
  SIZE_T        AddressSize,
  **DWORD**         dwMilliseconds
);

The page you've listed ([https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types](https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types)) says about DWORD:

DWORD **A 32-bit unsigned integer.** The range is 0 through 4294967295 decimal.
This type is declared in IntSafe.h as follows:
typedef unsigned long DWORD;

So it is 32 bit unsigned.

In C11 standard, chapter 7.8.1 Macros for format specifiers, it says: 

These macro names have the general form of PRI [...] followed by the conversion specifier, followed by a
name corresponding to a similar type name in 7.20.1. In these names, N represents the
width of the type as described in 7.20.1.

The fprintf macros for unsigned integers are: [...] PRIuN

Now in chapter in 7.20.1.1, it says: The typedef name uintN_t designates an unsigned integer type with width N and no
padding bits. 

I read uint32_t is an unsigned integer type with with 32 (bits) and no padding bits. Well - the range for that is exactly same as the range listed in windows-data-types. (0...2^32-1)

So I think that PRIu32 is a more appropiate format specified than %lu given the fact that the data type definition from the documentation says "32 bit unsigned integer". Which so happens that on this platform matches "unsigned long". 

I guess it all depends on what is canon here: "32 bit unsigned" or "typedef unsigned long DWORD". I tend to go the legalese way of "documentation trumps implementation". Anyway.

Will any of this (%PRIu32 vs %lu) change what is actually logged? :) Will it look different on the screen or in a log file?

(I also don't quite get the int32_t, fat finger'd twice?)

I guess I am going to leave this as is and talk some more at a later time.

---
In reply to: [423407935](https://github.com/Azure/azure-c-util/pull/37#discussion_r423407935) [](ancestors = 423407935,423407871)